torchdiffeq/_impl/misc.py

```diff
@@ -143,115 +146,67 @@ def _flat_to_shape(tensor, length, shapes):
     return tuple(tensor_list)
 
 
-class _TupleFunc(torch.nn.Module):
-    def __init__(self, base_func, shapes):
-        super(_TupleFunc, self).__init__()
-        self.base_func = base_func
-        self.shapes = shapes
+def _shape_to_flat(tensor):
+    return torch.cat([tensor_.reshape(-1) for tensor_ in tensor])
+
 
-    def forward(self, t, y):
-        f = self.base_func(t, _flat_to_shape(y, (), self.shapes))
-        return torch.cat([f_.reshape(-1) for f_ in f])
+_inf = torch.tensor(math.inf)
+_neginf = torch.tensor(-math.inf)
 
 
-class _ReverseFunc(torch.nn.Module):
-    def __init__(self, base_func):
-        super(_ReverseFunc, self).__init__()
+class _WrapFunc(torch.nn.Module):
+    def __init__(self, base_func, is_tuple, is_reversed, shapes):
+        super(_WrapFunc, self).__init__()
         self.base_func = base_func
+        self.is_tuple = is_tuple
+        self.is_reversed = is_reversed
+        self.shapes = shapes
 
-    def forward(self, t, y):
-        return -self.base_func(-t, y)
-
-
-def _check_inputs(func, y0, t, rtol, atol, method, options, SOLVERS):
```

Comment:
This got moved to avoid a circular import due to the RK check; if that's not the approach you're going for (see main comment) then indeed this may kept in misc.py

Updating func with a perturb argument is part of it. There's a few other things like adjusting (tuple/reversed) callbacks if necessary, and tidying how various options are handled.

Reply:
Ah, gotcha.